wepay/waltz
# PerformanceCli-client-creation-with-up-to-date-highwatermark

**D**escription:
If we want to run the ducktape tests back to back then one of the downside is the client should catch up to the latest transaction on the Waltz (while mounting) and this takes up time based on number of transactions written to Waltz. 
In order to avoid the waste of time (while mounting) and also to avoid resetting the cluster for every test, one option is to have the Waltz Client be able to start/mount from the latest transaction for that partition on Waltz. 

**S**olution:
Before multiple producer/consumer clients get created a simple client is created in _Map<Integer, Long> getHighWaterMark_. With the help of this client all watermarks for active partitions are stored in a HashMap. These values are populated to all clients in BaseCallbacks. HashMap has to be deep copied or otherwise all producer & consumer clients will share the same watermark.

**H**ow to test it:
Edit 
if (transaction.transactionId - 1 == partitionHighWaterMark.get(partition)) {
                    partitionHighWaterMark.put(partition, transaction.transactionId);
                    System.out.println("applying client clientHighWaterMark: " + partitionHighWaterMark.get(partition));
                }
to include the output line in producer/consumer apply transactions callback (to see later on that it actually works). 

performance_cli.py should include cmd_arr = [
            **"java -cp /usr/local/waltz/waltz-uber.jar",**
            "-Dlog4j.configuration=file:/etc/waltz-client/waltz-log4j.cfg", self.java_cli_class_name(),        

- build it _./gradlew clean build_ There is an extra test in PerformanceCliTest
- run performance tests in ducktape (will time out if number of clients and transactions isn't decreased - at least my experience - this is unrelated to this pull request)
- build waltz, scp the jar file to VMs (scp Waltz/waltz-uber/build/libs/waltz-uber-0.7.3-all.jar username@client-0:/usr/local/waltz
scp Waltz/waltz-uber/build/libs/waltz-uber-0.7.3-all.jar username@storage-0:/usr/local/waltz), update soft link to point out to this file in /usr/local/waltz, run the following command from a VM which serves as a client: java -cp /usr/local/waltz/waltz-uber.jar  -Dlog4j.configuration=file:/etc/waltz-client/waltz-log4j.cfg com.wepay.waltz.tools.performance.PerformanceCli test-consumers --txn-size 512 --num-txn 50 --cli-config-path /etc/waltz/waltz_client.yaml -uc --num-active-partitions 1 (delete -uc to see the difference)
